### ISSUE TITLE
deepstream@5.0.5 onClientDisconnect event not fired

### ISSUE THREAD

**Issue by caracal7** [deepstreamIO/deepstream.io#1038]
Server code
```
const { Deepstream } = require( '@deepstream/server' ); // v.5.0.5

const DeepstreamServer = new Deepstream( {
	httpServer: {
		type: 'default',
		options: {
			port: 6020,
			host: '127.0.0.1'
		}
	},
})

DeepstreamServer.set('authentication', {
	isValidUser( handshakeData, authData ) {
		return {
			isValid: true,
			id: authData.username,
		}
	},
	async whenReady () {
		console.log('whenReady')
	},
	//	This event never fired!!!
	onClientDisconnect( username ) {
		console.log( 'onClientDisconnect', username )
	},
});

DeepstreamServer.start();
console.log( 'DeepStream started...' );
```

**Comment by yasserf** [comment]
I think that might be because your not implementing the actual plugin interface.

Will look at it and get back to you

**Comment by caracal7** [comment]
```
const DeepstreamServer = new Deepstream( {
	httpServer: {
		type: 'default',
		options: {
			port: 6020,
			host: '127.0.0.1'
		}
	},
    auth: [{
        path: './PATH/my-authentication.js',
        options: {}
    }]
})
```

my-authentication.js

```
const { DeepstreamPlugin } = require('@deepstream/types')


module.exports = class TokenAuthentication extends DeepstreamPlugin {

    constructor ( pluginOptions,  services) {
         super()
    }

    async isValidUser(connectionData, authData) {
	 console.log('isValidUser')
         return null
    }

    async whenReady () {
	 console.log('whenReady')
    }

    //	This event never firead!!!
    async onClientDisconnect( username ) {
	console.log( 'onClientDisconnect', username )
    }

}
```

also didn't work

**Comment by caracal7** [comment]
@yasserf 
No ideas how to catch **onClientDisconnect** event?

**Comment by caracal7** [comment]
I didn't know TypeScript.
I make little hack witch solve my problem 

deepstream.io/src/connection-endpoint/base/connection-endpoint.ts
line:385
```
if (this.services.authentication.onClientDisconnect) {
  this.services.authentication.onClientDisconnect(socketWrapper.user)
```
I changed to 
```
// @ts-ignore
if (this.services.authentication.auths[0].onClientDisconnect) {
  // @ts-ignore
  this.services.authentication.auths[0].onClientDisconnect(socketWrapper.userId)
```
But this is not good solution 🤗

**Comment by yasserf** [comment]
Hey!

That’s almost completely correct, just needs to be a loop instead. I’ll put
in the fix. Thank you finding it!

**Comment by yasserf** [comment]
Should be fixed in 5.0.6, let me know!

**Comment by caracal7** [comment]
Version 5.0.7

deepstream.io/src/connection-endpoint/base/connection-endpoint.ts
line:386
```
this.services.authentication.onClientDisconnect(socketWrapper.user)
```
need to be replaced to
```
this.services.authentication.onClientDisconnect(socketWrapper.userId)
```
in order to **onClientDisconnect** return ID

**Comment by caracal7** [comment]
@yasserf 
Version 5.0.7

deepstream.io/src/connection-endpoint/base/connection-endpoint.ts
line:386
```
this.services.authentication.onClientDisconnect(socketWrapper.user)
```
need to be replaced to
```
this.services.authentication.onClientDisconnect(socketWrapper.userId)
```
in order to **onClientDisconnect** return ID

**Comment by yasserf** [comment]
Interesting that didn’t get caught by typescript. Will look at it.

**Comment by caracal7** [comment]
@yasserf did you fix with issue?

**Comment by yasserf** [comment]
I put in a fix but DS didnt release due an NPM issue which I didn’t look into

**Comment by caracal7** [comment]
Thanks a lot )

**Comment by yasserf** [comment]
Released in 5.0.9
